### PR TITLE
[org-node] Announce update to network peers after updating references

### PR DESCRIPTION
**Pull request by Ryanmtate**
Implements #56 

Builds off of #53

### CODE REVIEW

**Comment by Ryanmtate** [comment]
@cloudhead Think we're ready for review. the announce is only happening in the `track_peer_handler`, want to confirm that this is the only place we need it.

You had mentioned a startup process for updating the refs, which I think may still need to be added, but want to consult with you where it should be implemented.

File: org-node/src/client.rs [hunk 1]
```diff
@@ -359,12 +359,22 @@ impl Client {
...
+                            if let Err(err) = Client::sign_refs(urn.clone(), &api).await {
                                 tracing::error!(target: "org-node", "Error signing refs: {}", err);
                                 ack.send(Err(err)).ok();
                                 continue;
                             }
 
+                            // Announce the peer update to the network.
+                            if let Err(e) = api.announce(Payload {
+                                urn,
+                                // QUESTION: Should the current revision be added here?
```
Comment (cloudhead) [inline org-node/src/client.rs:371]: Yeah, I would say so. Don't quite understand what happens if you don't include a rev, ie. how peers are supposed to interpret that..

File: org-node/src/client.rs [hunk 2]
```diff
@@ -359,12 +359,22 @@ impl Client {
...
+                                rev: None,
+                                origin: Some(peer_id),
+                            }) {
+                                tracing::error!(target: "org-node", "Error announcing peer update: {:?}", e);
```
Comment (cloudhead) [inline org-node/src/client.rs:375]:
```suggestion
                                tracing::error!(target: "org-node", "Error announcing ref update: {:?}", e);
```

**Review by cloudhead — CHANGES_REQUESTED** [review]

**Comment by cloudhead** [comment]
> @cloudhead Think we're ready for review. the announce is only happening in the `track_peer_handler`, want to confirm that this is the only place we need it.
> 

Not sure I get the intent here. What exactly are we announcing to the network? We need to announce after a push has happened.

**Comment by Ryanmtate** [comment]
> > @cloudhead Think we're ready for review. the announce is only happening in the `track_peer_handler`, want to confirm that this is the only place we need it.
> 
> Not sure I get the intent here. What exactly are we announcing to the network? We need to announce after a push has happened.

Good point, I think I had a second announce on the Gossip applied update match arm, must have deleted it when handling the merge conflict by accident...

**Comment by Ryanmtate** [comment]
> > @cloudhead Think we're ready for review. the announce is only happening in the `track_peer_handler`, want to confirm that this is the only place we need it.
> 
> Not sure I get the intent here. What exactly are we announcing to the network? We need to announce after a push has happened.

The thinking with `track_peer_handler` is that it calls `set_head`, which I think we would want to announce to the network... unless I am mistaken 🤔

File: org-node/src/client.rs [hunk 3]
```diff
@@ -353,20 +353,32 @@ impl Client {
...
+                                        continue;
+                                    }
+
+                                    // Announce the peer update to the network after setting the head.
+                                    if let Err(e) = api.announce(Payload {
+                                        urn,
+                                        rev: Some(rev),
+                                        origin: Some(peer_id),
+                                    }) {
+                                        tracing::error!(target: "org-node", "Error announcing ref update: {:?}", e);
```
Comment (Ryanmtate) [inline org-node/src/client.rs:375]: also using debug print here due to error type of `Gossip::Payload`, can log an oid/urn instead if desired.

File: org-node/src/client.rs (hunk 1 above)
Comment (Ryanmtate) [reply org-node/src/client.rs]: removed announce statement

File: org-node/src/client.rs (hunk 3 above)
Comment (Ryanmtate) [reply org-node/src/client.rs]: removed announce statement from protocol events as extraneous

File: org-node/src/client.rs [hunk 4]
```diff
@@ -587,6 +587,21 @@ impl Client {
...
+
+                let oid = reference.target().expect("reference target must exist");
+
+                // Announce the updated refs to the network.
+                if let Err(e) = api.announce(Payload {
+                    urn,
+                    rev: Some(Rev::Git(oid)),
+                    origin: Some(api.peer_id()),
+                }) {
+                    tracing::error!(target: "org-node", "Error announcing refs: {:?}", e);
```
Comment (cloudhead) [inline org-node/src/client.rs:602]:
```suggestion
                    tracing::error!(target: "org-node", "Error announcing refs: {}", e);
```

File: org-node/src/client.rs [hunk 5]
```diff
@@ -587,6 +587,21 @@ impl Client {
                     )
                     .map_err(|e| Error::SetHead(Box::new(e)))?;
 
+                let reference = repository
+                    .find_reference(local_branch_ref.to_str().unwrap_or_default())
+                    .map_err(|e| Error::SetHead(Box::new(e)))?;
```
Comment (cloudhead) [inline org-node/src/client.rs:592]: Don't we already have the reference? Like, why are we trying to find it? Is it to make sure it exists?

Comment (Ryanmtate) [reply org-node/src/client.rs:592]: Can return from `reference_symbolic` and remove extraneous lookup.

Comment (Ryanmtate) [reply org-node/src/client.rs]: > Don't we already have the reference? Like, why are we trying to find it? Is it to make sure it exists?

Need to announce the `Rev::Git(oid)`

6fe4e73 removes additional lookup

File: org-node/src/client.rs (hunk 4 above)
Comment (Ryanmtate) [reply org-node/src/client.rs:598]: Error is payload type, so it does not implement display, maybe can use `e.urn.encode_id()` ?

Comment (cloudhead) [reply org-node/src/client.rs:598]: Oh weird, let's leave it like this for now.

**Review by cloudhead — APPROVED** [review]

**Comment by cloudhead** [comment]
Merged in 7a996be

**Pull request closed without merging**